### DISCUSSION TITLE
How to perform Wildcard, Fuzzy, Proximity or Fielded Searches in Pyserini?

### DISCUSSION THREAD

**Discussion by stefantrapp** [castorini/pyserini#1315] [Q&A]
Hi, when trying out the LuceneSearcher class on a simple example index like

{"id": "doc1", "contents": "contents of doc one."}
{"id": "doc2", "contents": "contents of document two."}
{"id": "doc3", "contents": "here's some text in document three."}

I couldn't find out how to use the full power of the Lucene Query Parsers in Pyserini. For example

hits = searcher.search("doc*") does only return the result

1 doc1 0.53460

i.e. it's not performing Wildcard Search as I expected. (How) Is it possible to do Wildcard Search, Fuzzy Search, Proximity Search and so on, or a Fielded Search like for example 

searcher.search("id:doc2")

Thanks.

**Comment by lintool** [comment]
Hi @stefantrapp - check out https://github.com/castorini/pyserini/blob/master/docs/usage-querybuilder.md

The entire Lucene query API should be available.

**Comment by stefantrapp** [comment]
Thanks, Jimmy, for your quick response and the link. I succeded to do a "Fielded Search" like this

tq = querybuilder.get_term_query('doc2', 'id')
hits = searcher.search(tq)

but unfortunately I'm still struggling with Wildcards or Proximity. Maybe the example you linked can be elaborated a bit more on this?


Sorry for following up. Thanks and kind regards,

Stefan

________________________________
Von: Jimmy Lin ***@***.***>
Gesendet: Samstag, 22. Oktober 2022 15:04:41
An: castorini/pyserini
Cc: Trapp, Stefan; Mention
Betreff: [Extern] Re: [castorini/pyserini] How to perform Wildcard, Fuzzy, Proximity or Fielded Searches in Pyserini? (Discussion #1315)


Hi @stefantrapp<https://github.com/stefantrapp> - check out https://github.com/castorini/pyserini/blob/master/docs/usage-querybuilder.md

The entire Lucene query API should be available.

**Comment by HAKSOAT** [comment]
Hi @stefantrapp 

There's a PR in the works for this: https://github.com/castorini/pyserini/pull/1330

When merged, you'll be able to do something like:

```
query = querybuilder.get_standard_query('doc*')
hits = searcher.search(query)
```

or work with nested logic like:

```
query = querybuilder.get_complex_phrase_query('"contents doc*"~3')
hits = searcher.search(query)
```

Does this address your use case?

**Comment by stefantrapp** [comment]
Hi Habeeb,


Thanks for your email. That looks promising, yes.


Thanks and best regards,

Stefan

________________________________
Von: Habeeb Shopeju ***@***.***>
Gesendet: Sonntag, 30. Oktober 2022 22:54:01
An: castorini/pyserini
Cc: Trapp, Stefan; Mention
Betreff: [Extern] Re: [castorini/pyserini] How to perform Wildcard, Fuzzy, Proximity or Fielded Searches in Pyserini? (Discussion #1315)


Hi @stefantrapp<https://github.com/stefantrapp>

There's a PR in the works for this: #1330<https://github.com/castorini/pyserini/pull/1330>

When merged, you'll be able to do something like:

query = querybuilder.get_standard_query('doc*')
hits = searcher.search(query)


or work with nested logic like:

query = querybuilder.get_complex_phrase_query('"contents doc*"~3')
hits = searcher.search(query)


Does this address your use case?

**Comment by St4r4x** [comment]
@body_pyserini_1315.txt

**Comment by GioiaZheng** [comment]
A small update for anyone landing here now: the higher-level `querybuilder` helper is still intentionally small, but `LuceneSearcher.search()` accepts either a query string or a Lucene `Query` object. For wildcard, fuzzy, and proximity searches, the most direct route is to use Lucene's query parser through Pyserini's Java bridge and pass the parsed query to `search`.

Since #1330 did not land, I would not rely on helper names such as `get_standard_query` in current Pyserini.

```python
from pyserini.analysis import get_lucene_analyzer
from pyserini.pyclass import autoclass
from pyserini.search.lucene import LuceneSearcher

searcher = LuceneSearcher("path/to/index")
analyzer = get_lucene_analyzer()

QueryParser = autoclass("org.apache.lucene.queryparser.classic.QueryParser")
parser = QueryParser("contents", analyzer)

# Fielded search
hits = searcher.search(parser.parse("id:doc2"), k=10)

# Wildcard
hits = searcher.search(parser.parse("contents:doc*"), k=10)

# Fuzzy
hits = searcher.search(parser.parse("contents:retrievel~1"), k=10)

# Proximity / slop
hits = searcher.search(parser.parse('"information retrieval"~3'), k=10)
```

For phrase/proximity queries that also need wildcard expansion inside the phrase, use Lucene's `ComplexPhraseQueryParser`:

```python
ComplexPhraseQueryParser = autoclass(
    "org.apache.lucene.queryparser.complexPhrase.ComplexPhraseQueryParser"
)
complex_parser = ComplexPhraseQueryParser("contents", analyzer)

hits = searcher.search(complex_parser.parse('"doc* retrieval"~3'), k=10)
```

One caveat: this bypasses the higher-level `querybuilder` helpers, so field names need to match how the index was built, and analyzer behavior still matters. If you only need a single analyzed term in a specific field, `querybuilder.get_term_query("doc2", field="id")` remains the simpler option.
